File: learning_parameter/learning_parameter/param_declare.py
import rclpy
from   rclpy.node import Node     # ROS2 节点类

class ParameterNode(Node):
    def __init__(self, name):
        super().__init__(name)
        timer_period = 2
        self.timer = self.create_timer(timer_period, self.timer_callback)

        self.declare_parameter('robot_name', 'mbot')

    def timer_callback(self):
        robot_name_param = self.get_parameter('robot_name').get_parameter_value().string_value

        self.get_logger().info('Hello %s!' % robot_name_param)

        new_name_param = rclpy.parameter.Parameter('robot_name', rclpy.Parameter.Type.STRING, 'mbot')
        all_new_parameters = [new_name_param]
        self.set_parameters(all_new_parameters)

def main():
    rclpy.init()
    node = ParameterNode('param_declare')
    rclpy.spin(node)
    node.destroy_node()                              # 销毁节点对象
    rclpy.shutdown()                                 # 关闭ROS2 Python接口

File: learning_parameter_cpp/src/param_declare.cpp
/***
@作者: 古月居(www.guyuehome.com)
@说明: ROS2参数示例-创建、读取、修改参数
***/

#include <chrono>
#include <functional>
#include <string>

#include <rclcpp/rclcpp.hpp>    // ROS2 C++接口库

using namespace std::chrono_literals;

class ParameterNode : public rclcpp::Node
{
    public:
        ParameterNode()
        : Node("param_declare")                                             // ROS2节点父类初始化
        {
            this->declare_parameter("robot_name", "mbot");

            timer_ = this->create_wall_timer(                               
                1000ms, std::bind(&ParameterNode::timer_callback, this));   // 创建一个定时器，定时执行回调函数
        }
         
        // 创建定时器周期执行的回调函数 
        void timer_callback()     
        {
            std::string robot_name_param = this->get_parameter("robot_name").as_string();                // 从ROS2系统中读取参数的值
            RCLCPP_INFO(this->get_logger(), "Hello %s!", robot_name_param.c_str());                      // 输出日志信息，打印读取到的参数值
            std::vector<rclcpp::Parameter> all_new_parameters{rclcpp::Parameter("robot_name", "mbot")};  // 重新将参数值设置为指定值
            this->set_parameters(all_new_parameters);                                                    // 将重新创建的参数列表发送给ROS2系统
        }

    private:
        rclcpp::TimerBase::SharedPtr timer_;
};

// ROS2节点主入口main函数
int main(int argc, char ** argv)                         
{

    // ROS2 C++接口初始化
    rclcpp::init(argc, argv);       
    
    // 循环等待ROS2退出                   
    rclcpp::spin(std::make_shared<ParameterNode>());   
    
    // 关闭ROS2 C++接口
    rclcpp::shutdown();                                
      
    return 0;
}
